### ISSUE TITLE
`Result.root` that is typed as `Document | Root` breaks the semantics of the `root` property

### ISSUE THREAD

**Issue by romainmenke** [postcss/postcss#1868]
Ever since the addition of `Document` I've been "arguing" with the types of PostCSS a lot more.
I've also been unable to add meaningful support for any syntax that makes use of the `Document` node because those syntax plugins have blocking issues that prevent real support.

Any plugin works equally bad or well as it did before this addition and none leverage this new node type.

So from my perspective the addition of `Document` isn't beneficial.

I feel that the current implementation isn't quite right.  
I don't know what should be changed to fix this.

-------

This specific issue is about this code :

https://github.com/postcss/postcss/blob/a0a82d3530ce7dfbfe1f12aa76c20f1b3f95eb7c/lib/result.d.ts#L144

What I want to do :

- parse a stylesheet that is a `foo.css` file
- access the `Root` node

I already know that it is a `Root` node but now the type system is complaining and saying that it can be a `Document` node.

I can override this and tell TypeScript that it really is a `Root` node.  
But what if someone uses one of the syntaxes that does return a `Document` in this case.

Which steps should I take :
- warn on `Document` nodes in `Result.root`?
- throw an error?
- traverse the `Document` nodes in search of `Root` nodes?

-------

I think `result.root` must always return a `Root` node.  
Never a `Document` node.

Maybe a new property can be added to expose `Document` nodes?

**Comment by ai** [comment]
We can add conditional types calculated from `opts.parser`

**Comment by romainmenke** [comment]
I don't think that would help because `result` is also exposed to plugins through `postcss.Helpers`

<img width="446" alt="Screenshot 2023-08-14 at 12 39 49" src="https://github.com/postcss/postcss/assets/11521496/cc05519b-427b-401b-aab0-a98e83eee07c">

**Comment by ai** [comment]
In plugins, you _must_ to be ready for `Document`. Otherwise, plugin will not work in any case.

But `result.root` is a very hacky and rare way.

**Comment by romainmenke** [comment]
Just to clarify, I am fine with this being unfixable :)

There is nothing here I can not work around, so this isn't blocking in any way.

I only wanted to give feedback on the impact of `Document`.

**Comment by romainmenke** [comment]
> In plugins, you must to be ready for Document. Otherwise, plugin will not work in any case.

🤔 But this conflicts a bit with the current guidance that this node type is experimental and might change.

> Experimental: some aspects of this node could change within minor or patch version releases.

So I've been trying very hard not to write code that is specific to `Document` because I don't want to run into issues with semver versioning.

**Comment by ai** [comment]
> But this conflicts a bit with the current guidance that this node type is experimental and might change.

Right now real users is using `Document`

> So I've been trying very hard not to write code that is specific to Document because I don't want to run into issues with semver versioning

For plugins, just use `Root`.

**Comment by romainmenke** [comment]
It's not possible to only use `Root` in plugins :)

```js
decl.parent?.parent
```

The parent of a `Container` can also be a `Document`.

<img width="730" alt="Screenshot 2023-08-14 at 19 56 53" src="https://github.com/postcss/postcss/assets/11521496/0f98d52c-d44e-433a-94b0-5c8fdf388c93">

The addition of `Document` does really add some complexity.

And things like a `result.root` that possible isn't a `Root` at all makes the PostCSS API a bit less ergonomic.

------------

As I said before, this is all fine and I can add more code to handel all the edge cases.
If the API is stable (although it is documented as experimental) then it's ok to leave this as is and to close this issue.

However if the API truly is still experimental and open to feedback then I think it would be beneficial to to review a few details.

**Comment by ai** [comment]
Nope, we are planning to use `Document` and don’t plan many changes.

**Comment by romainmenke** [comment]
Alright, thank you for taking the time to answer my questions here.

**Comment by ai** [comment]
I added `Result#root` casting from `process({ parser })` type 7d51976a

It will be `Root` if we didn’t pass `parser` option and will have `Document` type only if `opts.parser` types return `Document`.

By if we don’t know source of the `Result` (`{ result }` in plugin) it will continue to be `Root | Document`.

What do you think? Will it simplify PostCSS API for you?

**Comment by romainmenke** [comment]
I closed this because I thought no changes were going to be made for this.
Opening it again for visibility.

-----

That change will be helpful for end users and for creators of syntax plugins, thank you for that :)

But for plugins or runners the situation remains the same.

I think a more meaningful (but breaking) change would be to have a strict hierarchy and separation between `Document` and `Root`. This is why I asked if the API was still open to feedback. I am fairly certain that the situation for plugin authors can not be improved without breaking changes.

This is fine, it's natural that additions to the API add some complexity :)

-----

If we draw a parallel to HTML and CSS as it works in browsers we can observe several things:
- there is always a single document
- a document can load multiple stylesheets
- a stylesheet can never exist on it's own, it always belongs to a document
- a stylesheet can load other stylesheets
- a stylesheet can not load documents

This is a very strict and well defined hierarchical relationship.

In PostCSS a `Document` is similar to an browser document and `Root` is similar to a stylesheet. Similar but not exactly the same.

PostCSS doesn't make the distinction between a style declaration block that is the content of a `style` attribute in html and a stylesheet in a `<style>` element.

-------

If we didn't have to consider existing API and adoption of that API, it might be better to have something like :

- `Result` only has a `document` property, and has this in any syntax
- `Result` does not have a `root` property
- `Declaration.parent` can never be `Document`
- `Root.parent` is always a `Document`

For any syntax (like the default syntax) that doesn't have an actual document there would still be a `Document` node and it would contain a single `Root` node.

-------

I think that what is noticeable here is that it is much easier to add a new AST node further down the AST than it is to add a new container around everything.

And again let me say that this is fine.
I don't expect everything to be fixed :)

But I still wanted to give feedback on some of things I encounter.

**Comment by ai** [comment]
Unfortunately, it will be to big change for small use case.

Where exactly for you have a problem with `result.root` case? Maybe I can suggest a better way to handle it if you will provide me more context.

**Comment by romainmenke** [comment]
For `result.root` I am now using recursive functions.

```ts
function doSomethingWithAResult(styles: Root | Document) {
 	if (styles.type === 'document') {
 		styles.each((root) => {
 			doSomethingWithAResult(root);
 		});

 		return;
 	}

 	// do real work here
}
```

This works well enough, but it feels like unneeded complexity.

I wouldn't write the code like this if the type of `result.root` was only `Document` or only `Root`.

If they were separate I would write a dedicated function that takes `Document` typed arguments and one for `Root` typed arguments.

Even if it is just this :

```ts
function doSomethingWithADocument(document: Document) {
 	document.each((root) => {
 		doSomethingWithARoot(root);
 	});
}

function doSomethingWithARoot(root: Root) {
 	// do real work here
}
```

-----

But again, I don't expect a solution here :)
I know various patterns to get the desired result and these are fine.

There is just a bit more friction than there used to be before `Document` was introduced.

**Comment by ai** [comment]
The question is where do you use `doSomethingWithADocument`. Maybe you can give me a link to real code?

**Comment by romainmenke** [comment]
Currently working on a stricter bundler :

First path :

1. https://github.com/csstools/postcss-plugins/blob/dee07e14cd1abcff20eb2b80d8015c3b958f8b57/plugin-packs/postcss-bundler/src/postcss-import/index.ts#L13
2. https://github.com/csstools/postcss-plugins/blob/dee07e14cd1abcff20eb2b80d8015c3b958f8b57/plugin-packs/postcss-bundler/src/postcss-import/lib/parse-styles.ts#L18
3. https://github.com/csstools/postcss-plugins/blob/dee07e14cd1abcff20eb2b80d8015c3b958f8b57/plugin-packs/postcss-bundler/src/postcss-import/lib/parse-statements.ts#L6

Second path :
1. https://github.com/csstools/postcss-plugins/blob/dee07e14cd1abcff20eb2b80d8015c3b958f8b57/plugin-packs/postcss-bundler/src/postcss-import/lib/parse-styles.ts#L147
2. https://github.com/csstools/postcss-plugins/blob/dee07e14cd1abcff20eb2b80d8015c3b958f8b57/plugin-packs/postcss-bundler/src/postcss-import/lib/parse-styles.ts#L18
3. https://github.com/csstools/postcss-plugins/blob/dee07e14cd1abcff20eb2b80d8015c3b958f8b57/plugin-packs/postcss-bundler/src/postcss-import/lib/parse-statements.ts#L6

The first path starts in `Once` where `root` is of type `Root`.

The second path starts with loading a new file and building the AST, here the `root` can be a `Document`.

In both cases the syntax/parser that is used is unknown as it is user provided.

-------

`postcss-import` itself doesn't use TypeScript and would just error if a `Document` node was returned. So using TypeScript exposed a bug in that implementation.

**Comment by ai** [comment]
Yes, in `Once` there is no other way 😅

Sorry, it is not very popular to fix (and I don’t know simple solution).

I will release fix to `process().root` and close the issue.

**Comment by ai** [comment]
The `process().root` fix was released in 8.4.28.
